Custom JWT plugin (& RelayState small fix)

sgabb opened the pull request:
A fix and a feature:
- RelayState can also be a "/" (%2F char), so I added an if check in order to not evaluate the slash as a token
- Add in settings a trigger to pass a custom function to create token query instead of using PySAML token

mostafa commented on README.md:
```diff
@@ -177,6 +177,7 @@ For IdP-initiated SSO, the user will be created if it doesn't exist, but for SP-
 | **TRIGGER.BEFORE\_LOGIN**                   | A method to be called when an existing user logs in. This method will be called before the user is logged in and after the SAML2 identity provider returns user attributes. This method should accept ONE parameter of user dict.                                                                                                                               ...
 | **TRIGGER.AFTER\_LOGIN**                    | A method to be called when an existing user logs in. This method will be called after the user is logged in and after the SAML2 identity provider returns user attributes. This method should accept TWO parameters of session and user dict.                                                                                                                   ...
 | **TRIGGER.GET\_METADATA\_AUTO\_CONF\_URLS** | A hook function that returns a list of metadata Autoconf URLs. This can override the `METADATA_AUTO_CONF_URL` to enumerate all existing metadata autoconf URLs.                                                                                                                                                                                                 ...
+| **TRIGGER.CUSTOM\_TOKEN\_QUERY** | A method to be called to create a custom JWT for the user. This method will be called between `BEFORE_LOGIN` and `AFTER_LOGIN` triggers and should return the query params containing the token (e.g., `?token=jwt.token.encripted`). This method should accept ONE parameter of user.                                                                                     ...
```
```suggestion
| **TRIGGER.CUSTOM\_TOKEN\_QUERY** | A hook function to create a custom JWT for the user. This method will be called between `BEFORE_LOGIN` and `AFTER_LOGIN` triggers and should return the query params containing the token (e.g., `?token=encoded.jwt.token`). This method accepts one parameter: `user`.                                                                                                                                                                                                                                                                                             | `str`            | `None`                                                                                                                                   | `my_app.models.users.get_custom_token_query`         |
```

mostafa commented on django_saml2_auth/views.py:
```diff
@@ -73,8 +73,9 @@ def acs(request: HttpRequest):
+    # RelayState, depending on the SAML app configuration, can also be a "/" (%2F char)
     relay_state = request.POST.get("RelayState")
-    if relay_state:
+    if relay_state and relay_state != '/':
```
I think this must be adequately validated. Just checking if it exists or is not equal to something is not proper validation. WDYT?

sgabb replied:
What about something like this:

```python
def is_jwt_well_formed(jwt: str):
    """Check if JWT is well formed

    Args:
        jwt (str): Json Web Token

    Returns:
        Boolean: True is JWT is well formed, otherwise False
    """
    if isinstance(jwt, str):
        # JWT should contain three segments, separated by two period ('.') characters.
        jwt_segments = jwt.split('.')
        if len(jwt_segments) == 3:
            jose_header = jwt_segments[0]
            # base64-encoded string length should be a multiple of 4
            if len(jose_header) % 4 == 0:
                try:
                    jh_decoded = base64.b64decode(jose_header).decode('utf-8')
                    if jh_decoded and jh_decoded.find('JWT') > -1:
                        return True
                except Exception:
                    return False
    # If tests not passed return False
    return False

token = 'eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJzdWIiOiI0MjQyIiwibmFtZSI6Ikplc3NpY2EgVGVtcG9yYWwiLCJuaWNrbmFtZSI6Ikplc3MifQ.EDkUUxaM439gWLsQ8a8mJWIvQtgZe0et3O3z4Fd_J8o'
res = is_jwt_well_formed(jwt=token)
print(res)  # True
res = is_jwt_well_formed(jwt='/')
print(res)  # False
```

and then in views
```python
relay_state_is_token = is_jwt_well_formed(relay_state)
    if relay_state_is_token:
```

mostafa replied:
There's no need to do this manually, as it might be error-prone. Decoding a JWT token actually validates it, too, in addition to checking all the fields, as documented [here](https://pyjwt.readthedocs.io/en/stable/usage.html#registered-claim-names).

sgabb replied:
The idea was not to check the token manually, but to check wether the relay state could be a token to check.
The RelayState is not necessary a token (as stated [here](https://developer.okta.com/docs/concepts/saml/#single-idp-vs-multiple-idps) and [there](https://stackoverflow.com/a/34351756)), but can be whatever.
So it seemed better to check briefly wether to enter the block where it actually checks if the token is valid, otherwise we would not really know if the token is not encoded correctly or is simply not a token to be checked.

mostafa replied:
Fair enough!

mostafa commented on django_saml2_auth/views.py:
```diff
@@ -97,9 +98,16 @@ def acs(request: HttpRequest):
+        # Check if there is a custom trigger for creating the JWT and URL query
+        custom_jwt_query_trigger = dictor(saml2_auth_settings, "TRIGGER.CUSTOM_TOKEN_QUERY")
+        if custom_jwt_query_trigger:
+            query = run_hook(custom_jwt_query_trigger, target_user)
```
How about [decoding](https://github.com/grafana/django-saml2-auth/blob/9213beeabc17e0990c6c08d15e8efa450bc5fff1/django_saml2_auth/views.py#L137)?

sgabb replied:
Actually I wasn't using the sp_initiated_login view but only "signin" and "acs"; what should be the use case of the sp initiated login? I mean, where does the token in the GET method come from?

mostafa replied:
It usually comes from a redirect in your webapp.

sgabb replied:
So, at the moment the flow I am using is the following:
- frontend webapp calls the backend django rest api at django-saml2-auth "signin" view
- backend creates an AuthNRequest and redirect to IdP
- user login at the IdP and is redirected to backend
- backend confirm AuthNResponse, creates a JWT token and redirects the user to the frontend webapp with the token to be used in the following api backend calls

In the sp_initiated_login the case could be that the frontend webapp is calling the login with a token to be checked or refreshed by the IdP?

However, probably to make it more customisable we could add another trigger function for checking and refreshing the token?
In the view it gets the user_id from the token, prepare saml, recreate a token and pass it to the RelayState:
```python
user_id = decode_jwt_token(request.GET.get("token"))
saml_client = get_saml_client(get_assertion_url(request), acs, user_id)
jwt_token = create_jwt_token(user_id)
```

How do you think would be a good way to do that hook?

mostafa replied:
@sgabb 

In the SP-initiated login, this is the flow:
- Frontend sends a login request to the backend (usually just an email).
- Backend verifies the requests and returns a redirect URL with a signed JWT token to the frontend.
- Frontend redirects the user to the URL along with the token. The URL points to the sp_initiated login endpoint.
- The endpoint refreshes the token and then the user is redirected to the IdP login page with the new token in their AuthNRequest.
- After login, the flow is the same as the IdP-initiated login that goes through the `/sso/acs` endpoint.

I suggest that if we want to decode the token somewhere, just like encoding, we should consider having another hook that deals with decoding custom tokens. The `create_jwt_token` could make use of the `TRIGGER.CUSTOM_TOKEN_QUERY` hook and probably be renamed to something more meaningful, like `CUSTOM_CREATE_JWT_TOKEN`, and then the `decode_jwt_token` hook can use another hook with the reverse logic, called `CUSTOM_DECODE_JWT_TOKEN`, which of course, should be documented. The names are just suggestions, so feel free to use any name that makes more sense.

sgabb replied:
@mostafa 
I am working on it.
Quick question: how should I run the test in development?
I tried just running `pytest` but I get _"No module named 'django_saml2_auth.tests.settings'"_ :sweat:

sgabb commented:
So, I refactored the code to use the trigger custom function to create and decode JWT in all the library.
Changed the test so that they use the "custom_or_default_jwt" functions and everything is passing :partying_face: 
Updated readme descriptions and examples.

mostafa commented on README.md:
```diff
@@ -177,6 +177,9 @@ For IdP-initiated SSO, the user will be created if it doesn't exist, but for SP-
 | **TRIGGER.GET\_METADATA\_AUTO\_CONF\_URLS** | A hook function that returns a list of metadata Autoconf URLs. This can override the `METADATA_AUTO_CONF_URL` to enumerate all existing metadata autoconf URLs.                                                                                                                                                                                                 ...
+| **TRIGGER.CUSTOM\_DECODE\_JWT** | A hook function to decode the user JWT. This method will be called instead of the `decode_jwt_token` default function and should return the user_model.USERNAME_FIELD. This method accepts one parameter: `token`.                                                                                                                                                          ...
+| **TRIGGER.CUSTOM\_CREATE\_JWT** | A hook function to create a custom JWT for the user. This method will be called instead of the `create_jwt_token` default function and should return the token. This method accepts one parameter: `user`.                                                                                                                                                                  ...
+| **TRIGGER.CUSTOM\_TOKEN\_QUERY** | A hook function to create a custom query params with the jwt for the user. This method will be called after `CUSTOM_CREATE_JWT` to populate a query and attach it to a URL; should return the query params containing the token (e.g., `?token=encoded.jwt.token`). This method accepts one parameter: `token`.                                                            ...
```
Two things:
1. After adding the other two custom hooks, the `TRIGGER.CUSTOM_TOKEN_QUERY` hook seems like it can be narrowed down to just the token parameter as a constant in settings, rather than a custom hook. Or if you have something more complex on your mind, please update the example to include that.

2. Nitpick:
```suggestion
| **TRIGGER.CUSTOM\_TOKEN\_QUERY** | A hook function to create a custom query params with the JWT for the user. This method will be called after `CUSTOM_CREATE_JWT` to populate a query and attach it to a URL; should return the query params containing the token (e.g., `?token=encoded.jwt.token`). This method accepts one parameter: `token`.                                                                                                                                                                                                                                                                                             | `str`            | `None`                                                                                                                                   | `my_app.models.users.get_custom_token_query`         |
```

sgabb replied:
1. The custom token query could be useful if the query is different than just `?token_name=token`, for example with access and refresh tokens. I have written an example in the [readme](https://github.com/grafana/django-saml2-auth/pull/35/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R246-R264) with a query function that get the token and create a query by using other library features and create an `?access=access.token&refresh=refresh.token` query.
2. Accepted!

mostafa reviewed (commented):
Well-done! 🚀
I just have a question and a nit.

mostafa commented:
@sgabb Thanks for your contribution. I'll merge this PR.
